### ISSUE TITLE
User interface for masking output on `ImmersedBoundaryGrid`

### ISSUE THREAD

**Issue by glwagner** [CliMA/Oceananigans.jl#3061]
I have found it inconvenient that output is masked to `0`. For most purposes, I'd prefer `NaN`. I'm opening this issue to discuss a user interface that would implement such a feature.

My first idea is to add a property / kwarg to output writers called `immersed_values`.

Then we might write something like

```julia
output_writer = JLD2OutputWriter(model, output, immersed_values=NaN, other_kwargs...)
```

The default will be `immersed_values=nothing`. Then `fetch_output`:

https://github.com/CliMA/Oceananigans.jl/blob/72e2197e640ddf308fd40aa4658e7686e596bccd/src/OutputWriters/fetch_output.jl#L12-L15

would become something like

```julia

 function fetch_output(field::AbstractField, model, immersed_value) 
    compute_at!(field, time(model))
    !isnothing(immersed_value) && mask_immersed_field!(field, immersed_value)
    return parent(field) 
 end 
```

**Comment by glwagner** [comment]
@tomchor

**Comment by navidcy** [comment]
Yeah, let's do that! I'll open a PR

**Comment by glwagner** [comment]
What kwarg name will you use?

**Comment by glwagner** [comment]
Let’s discuss the user API here though it’s pretty simple

**Comment by navidcy** [comment]
you mean what kwarg for the `JLD2/NetCDFOutputWriter`? perhaps `replace_land_with_NaN = true` ?

**Comment by glwagner** [comment]
Mm yeah that’s an alternative! I think we could allow any number. As for “land” versus “immersed” I guess land is a little more oceany but elsewhere we talk about immersed stufff…

**Comment by glwagner** [comment]
Or immersed_mask = 0?

**Comment by glwagner** [comment]
The concept of “fill” or “replace” might be helpful. Like “fill_immersed=0”

**Comment by simone-silvestri** [comment]
I like the concept of `mask`, what about `immersed_mask_value=0`?

**Comment by glwagner** [comment]
`mask_immersed=0`

**Comment by navidcy** [comment]
I vote for simply `mask = default_value` and `mask!(field)`

**Comment by navidcy** [comment]
The `mask_value` should be added as a field of each writer type, right?

**Comment by glwagner** [comment]
yes, it has to be a property of JLD2OutputWriter and NetCDFOutputWriter

**Comment by navidcy** [comment]
```
mask_immersed = 0
```

and

```
mask_immersed!(field)
```
?

**Comment by glwagner** [comment]
I like it! The concept a function name that is something like `take_action_on(argument)` is typical, eg the function `fill!(array, value)` rather than `fill_value!(array, value)` (which is redundant for the reader). I liek the readability of `mask_immersed = 0`, as in "mask immersed [grid cells with] = value".

**Comment by tomchor** [comment]
Are we currently masking anything on outputs? There are no instances of the word `mask` in `OutputWriters`, so I'm thinking no:
```
$ grep -r "mask" --include "*jl"
$
```

In any case, discussing with @ali-ramadhan we realized that masking is also important for compression algorithms. So we should probably be doing this. Masking immersed regions with a constant value probably allows for a pretty significant file size reduction even with very conservative compression levels.

**Comment by ali-ramadhan** [comment]
> For compression, we want to mask to 0 right?

As long as it's the same value/bits at each masked grid point I think even basic compression algorithms will do a good job. Zero seems like the most natural choice, but we could make it an option in case someone wants to mask with `NaN` or `-1` or something else.

**Comment by glwagner** [comment]
That makes sense. NaN is helpful for plotting. But I think computations are slow with NaN. I like the idea of prioritizing post-processing and leaving visualization issues to plotting extensions (at least by default).

**Comment by tomchor** [comment]
> Note that the Makie plotting extension inserts NaN automatically. This negaties the motivation for masking output for visualization purposes (one reason this PR hasn't moved much)...

I don't think it _negates_ it since that's only true if you're using the Makie Ext, which isn't 100% of people. Even in Julia there are other plotting libraries that people might want to use.

**Comment by tomchor** [comment]
> Output should be masked to 0 currently I think because `mask_immersed_field!` is called during time-stepping. But someday, we should call it only for output.

Ah, I see. @ali-ramadhan and I were discussing why we didn't see much size difference when trying out different compression levels, but I guess our expectations were just too high :)

**Comment by glwagner** [comment]
Can you try it on some test data?
